x-pack/plugins/data_enhanced/server/search/es_search_strategy.ts

```diff
@@ -4,97 +4,77 @@
-import { getDefaultSearchParams, getAsyncOptions } from './get_default_search_params';
-
-import type { SharedGlobalConfig, Logger } from '../../../../../src/core/server';
-
 import type {
+  IEsSearchRequest,
+  IEsSearchResponse,
+  ISearchOptions,
   ISearchStrategy,
+  SearchStrategyDependencies,
   SearchUsage,
-  IEsRawSearchResponse,
-  ISearchOptions,
-  IEsSearchResponse,
 } from '../../../../../src/plugins/data/server';
-
-import type { IEnhancedEsSearchRequest } from '../../common';
-
-const { utils } = search.esSearch;
-
-interface IEsRawAsyncSearchResponse<Source = any> extends IEsRawSearchResponse<Source> {
-  response: SearchResponse<Source>;
-}
+import {
+  getDefaultSearchParams,
+  getShardTimeout,
+  getTotalLoaded,
+  searchUsageObserver,
+  shimAbortSignal,
+} from '../../../../../src/plugins/data/server';
+import type { IAsyncSearchOptions } from '../../common';
+import { pollSearch } from '../../common';
+import {
+  getDefaultAsyncGetParams,
+  getDefaultAsyncSubmitParams,
+  getIgnoreThrottled,
+} from './request_utils';
+import { toKibanaSearchResponse } from './response_utils';
 
 export const enhancedEsSearchStrategyProvider = (
   config$: Observable<SharedGlobalConfig>,
   logger: Logger,
   usage?: SearchUsage
 ): ISearchStrategy<IEnhancedEsSearchRequest> => {
   function asyncSearch(
-    request: IEnhancedEsSearchRequest,
-    options: ISearchOptions,
+    { id, ...request }: IEsSearchRequest,
+    options: IAsyncSearchOptions,
     { esClient, uiSettingsClient }: SearchStrategyDependencies
   ) {
-    const asyncOptions = getAsyncOptions();
     const client = esClient.asCurrentUser.asyncSearch;
 
-    return doPartialSearch<ApiResponse<IEsRawAsyncSearchResponse>>(
-      async () =>
-        client.submit(
-          utils.toSnakeCase({
-            ...(await getDefaultSearchParams(uiSettingsClient)),
-            batchedReduceSize: 64,
-            ...asyncOptions,
-            ...request.params,
-          })
-        ),
-      (id) =>
-        client.get({
-          id: id!,
-          ...utils.toSnakeCase({ ...asyncOptions }),
-        }),
-      (response) => !response.body.is_running,
-      (response) => response.body.id,
-      request.id,
-      options
-    ).pipe(
-      utils.toKibanaSearchResponse(),
-      map((response) => ({
-        ...response,
-        rawResponse: shimHitsTotal(response.rawResponse.response!),
-      })),
-      utils.trackSearchStatus(logger, usage),
-      utils.includeTotalLoaded()
+    const search = async () => {
+      const params = id
+        ? { ...getDefaultAsyncGetParams(), id }
+        : { ...(await getDefaultAsyncSubmitParams(uiSettingsClient)), ...request.params };
+      const promise = id ? client.get(params) : client.submit(params);
+      const { body } = await shimAbortSignal(promise, options.abortSignal);
+      return toKibanaSearchResponse(body);
+    };
+
+    return pollSearch(search, options).pipe(
+      tap((response) => (id = response.id)),
```

Comment:
Interesting way of passing the `id` back into the next call of search.
Might be worth a comment.